src/V3Number.h

```diff
@@ -56,69 +59,382 @@ class V3NumberData final {
+            // (Re)allocation with realloc allows for:
+            // - Increasing the buffer size without data loss when current size is not explicitly
+            //   known.
+            // - A chance of buffer resizing without moving it.
+            void* const ptr = std::realloc(oldPtr, sizeof(ValueAndX) * dynamicWordsCount);
+            m_dynamic.reset(reinterpret_cast<ValueAndX*>(ptr));
+        }
     }
+
     ValueAndX& operator[](size_t idx) {
-        UDEBUGONLY(UASSERT(idx < m_data.size() + m_inlinedSize,
-                           "idx:" << idx << " size:" << m_data.size()
-                                  << " inlinedSize:" << m_inlinedSize););
-        return idx >= m_inlinedSize ? m_data[idx - m_inlinedSize] : m_inlined[idx];
+        return idx >= m_inlinedSize ? m_dynamic[idx - m_inlinedSize] : m_inlined[idx];
+    }
+    const ValueAndX& operator[](size_t idx) const {
+        return const_cast<V3NumberFourStateArray&>(*this)[idx];
     }
-    const ValueAndX& operator[](size_t idx) const { return const_cast<V3NumberData&>(*this)[idx]; }
 };
 
-class V3Number final {
-    // TYPES
-    using ValueAndX = V3NumberData::ValueAndX;
+enum class V3NumberDataType : uint8_t {
+    UNINITIALIZED = 0,
+    STRING = 1,
+    DOUBLE = 2,
+    LOGIC = 3,
+};
+inline std::ostream& operator<<(std::ostream& os, const V3NumberDataType& rhs) {
+    switch (rhs) {
+    case V3NumberDataType::UNINITIALIZED: return os << "UNINITIALIZED";
+    case V3NumberDataType::STRING: return os << "STRING";
+    case V3NumberDataType::DOUBLE: return os << "DOUBLE";
+    case V3NumberDataType::LOGIC: return os << "LOGIC";
+    }
+    return os;
+}
+
+class V3NumberData final {
+    // MEMBERS
+    union {
+        uint8_t m_uninitializedMemberDoNotUse;  // Dummy member to avoid automatic initialization.
+        V3NumberFourStateArray m_number;
+        std::string m_string;
+    };
+
+    union {
+        FileLine* m_fileline;
+        AstNode* m_nodep;  // Parent node
+    };
 
-    // Large 4-state number handling
-    int m_width;  // Width as specified/calculated.
+    int m_width = 0;  // Width (in bits) as specified/calculated
+    V3NumberDataType m_type : 2;
+
+public:
     bool m_sized : 1;  // True if the user specified the width, else we track it.
     bool m_signed : 1;  // True if signed value
-    bool m_double : 1;  // True if double real value
     bool m_isNull : 1;  // True if "null" versus normal 0
-    bool m_isString : 1;  // True if string
     bool m_fromString : 1;  // True if from string literal
     bool m_autoExtend : 1;  // True if SystemVerilog extend-to-any-width
-    FileLine* m_fileline;
-    AstNode* m_nodep;  // Parent node
-    V3NumberData m_value;  // Value and X/Z information
-    string m_stringVal;  // If isString, the value of the string
+
+private:
+    bool m_hasNodep : 1;  // True if m_nodep is in use. False if m_fileline is.
+
+public:
+    // CONSTRUCTORS
+    V3NumberData()
+        : m_fileline(nullptr)
```

Comment:
Please use C++11 {} constructors in all new code thanks.

Reply:
Fixed